### DISCUSSION TITLE
Best Practice: Build in docker container or COPY?

### DISCUSSION THREAD

**Discussion by mmlngl** [nrwl/nx#29001] [Questions]
Hi there,

I’m working on a project where I use NX to manage a monorepo with a Remix app (`web`) and a library (`design-system`) that it depends on. I'm deploying the app using Docker and hosting it on Fly.io, and I’ve been exploring the best way to handle the build process for this setup.

## Context

Monorepo Structure:
- `apps/web`: The main Remix app.
- `libs/design-system`: A shared library that `web` depends on.
	
Build Process: The app needs to be built with the library’s code included, so both the app and its dependencies are fully compiled within the container.

The main challenge I've come-up against, while create a Docker build process is Dependency Management: Ensuring that web has access to the latest build of design-system within the Docker container.

# What I've Tried:

1. Building Locally and Copying Artifacts: Attempted building the app locally and copying the build output into the Docker container. However, dependencies for `design-system` were missing when I ran the container and I was worried about how particular binaries might work.

2. Multi-Stage Docker Builds: Bringing the entire workspace into the container and using nx commands to build and run the app. This felt bloated and the build step of the Remix app failed.

# Questions

- Have I been barking-up the wrong tree?
- What's the best practice here?

When I find an answer, I'll publish an example repo so nx neebies like me can learn too.

**Comment by yanniznik** [comment]
@mmmoli same question, have you figured out a proper way to do this?

**Reply by mmlngl** [reply]
Unfortunately no. I gave up and moved away from a docker set up. 

Best thing I found was `PNPM deploy`. I was playing with different package managers but deploy really changed things. 

If you run it, it creates a directory with only the dependencies that app needs. 

You can then copy that into docker and build from there. 

Pretty neat. 

---

In case you're interested, I'm running the same stack of Cloudflare workers now and I absolutely love the set up. Can't recommend it enough.
